### ISSUE TITLE
GPU decoder forward returns a silently wrong perplexity (49,494 vs 15.53 on CPU, same binary and weights)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#2084]
## Summary

Loading SmolLM2-360M through `PretrainedLoader` and running a dense forward gives a
**perplexity of 49,494.9 on the GPU path against 15.5346 on CPU** — same binary, same
weights, same window, same configuration. The only difference is which engine the
process selected at startup.

This is a silent wrong-answer, not a crash: the run completes, reports a number, and
nothing indicates the result is meaningless.

## Reproduction

Same executable, same arguments, twice. The only variable is whether the GPU is
available/enabled at process start.

```
# GPU selected
[Engine] DirectGpu available: CUDA (NVIDIA GeForce RTX 3080)
[Engine] DirectGpu: 70 CUs, 12.0GB VRAM
[mem after load] ws=8.73GB
[evalservice] densePpl=49494.90577568222   inter=2560 layers=32

# CPU (AIDOTNET_DISABLE_GPU=1)
[Engine] DirectGpu: Not available (no compatible GPU backends found)
[Engine] Auto-selected: CPU Engine
[mem after load] ws=10.73GB
[evalservice] densePpl=15.534647357808888  inter=2560 layers=32
```

The CPU value is reproducible to all 15 significant digits across separate process
launches, so the baseline itself is deterministic and the GPU value is the outlier.

Model: SmolLM2-360M, S=1024, hidden=960, inter=2560, V=49152, layers=32. Perplexity is
measured over one held-out 1024-token window.

## Scope: NOT a general GPU defect

A plain supervised path is unaffected. Building `MultipleRegression` through the facade
and predicting gives `5.0000` against a truth of `5.0` on **both** CPU and GPU. So the
divergence is specific to the pretrained-decoder forward, not to the engine generally.

## Notes toward a cause (unverified)

Two candidates, neither confirmed — recorded so whoever picks this up does not repeat
the search:

1. The RoPE half-split weight-permute fix **is present** in this branch
   (`LlamaModelBuilder.ShouldPermuteRope` / `permuteRopeQK`), so this is not simply a
   branch missing that work.
2. The 2 GB difference in working set after load (8.73 GB GPU vs 10.73 GB CPU) suggests
   the two paths do not materialise the same state. Worth checking whether the model is
   left in training mode after `PretrainedLoader.Load` — dropout active on one engine and
   not the other would produce exactly this shape of failure.

## Why this matters for v1

A silent numerical divergence between engines undermines every measurement taken on the
GPU path, and there is currently nothing that would catch it: a parity test comparing a
pretrained decoder forward across engines does not exist.

## Suggested acceptance criteria

- A test loads a small pretrained decoder, runs a fixed-window forward on CPU and on GPU,
  and asserts the perplexities agree within a stated tolerance.
- The test runs wherever a GPU is available and skips explicitly (not silently) elsewhere.
- Root cause identified and fixed, or the GPU decoder path documented as unsupported and
  made to fail loudly rather than return a number.

**Comment by ooples** [comment]
Closing this: **the defect is not in AiDotNet.** I filed it before verifying which component
produced the wrong number, and the measurement now shows AiDotNet's decoder forward agrees
across engines.

## What I actually measured

Loading the same SmolLM2-360M weights through `PretrainedLoader<float>.Load(...)` and calling
`net.Predict(...)` directly, checksumming the logits, once per engine:

| sequence length | GPU | CPU | relative difference |
|---|---|---|---|
| 32   | -24986.707    | -24987.617    | 0.0036%        |
| 128  | 7994592.5     | 7994597       | 0.00006%       |
| 512  | 53511784      | 53511784      | **bit-identical** |
| 1024 | 123352310     | 123352340     | 0.00002%       |

That is ordinary float32 accumulation-order variance between two engines, and it does not grow
with depth or sequence length. There is no corruption in this path.

## Where the 49,494 actually came from

The process that produced `densePpl=49494.9` does not call `Predict`. It is a private research
harness that reconstructs the transformer block forward itself from `PreLNTransformerBlock`'s
public sub-layer accessors (norm → attention → norm → gate/up/down) in order to capture
per-layer activations and swap the FFN. The divergence is in that reconstruction, not in the
library's own forward, and it is being pursued there.

## Two things worth keeping from this

- The training-mode hypothesis in the original report is also **disproved**:
  `PretrainedLoader` does leave `IsTrainingMode = true` (confirmed), but the forward is
  deterministic across repeated identical input and gives an identical checksum before and
  after `SetTrainingMode(false)`. No dropout is active in this stack, so the flag has no
  numerical effect here. Whether the loader *should* return an inference-mode model is a
  separate API question, not a correctness one.
- A cross-engine parity test for a pretrained decoder forward still does not exist in this
  repository. The numbers above are the baseline such a test would assert, and adding one is
  worth doing on its own merit — it is what would have located this in minutes instead of
  sending me to the wrong repository.

Apologies for the noise; I should have isolated the component before opening this.
